File: src/brs/unconfirmedtransactions/UnconfirmedTransactionStoreImpl.java
package brs.unconfirmedtransactions;

import brs.BurstException.ValidationException;
import brs.Constants;
import brs.Transaction;
import brs.db.store.AccountStore;
import brs.peer.Peer;
import brs.props.PropertyService;
import brs.props.Props;
import brs.services.TimeService;
import brs.transactionduplicates.TransactionDuplicatesCheckerImpl;
import brs.transactionduplicates.TransactionDuplicationResult;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.List;
import java.util.SortedMap;
import java.util.TreeMap;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import java.util.function.Consumer;
import java.util.stream.Collectors;
import org.apache.commons.lang.StringUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class UnconfirmedTransactionStoreImpl implements UnconfirmedTransactionStore {

  private static final Logger logger = LoggerFactory.getLogger(UnconfirmedTransactionStoreImpl.class);

  private final TimeService timeService;
  private final ReservedBalanceCache reservedBalanceCache;
  private final TransactionDuplicatesCheckerImpl transactionDuplicatesChecker = new TransactionDuplicatesCheckerImpl();

  private HashMap<Transaction, HashSet<Peer>> fingerPrintsOverview = new HashMap<>();

  private final SortedMap<Long, List<Transaction>> internalStore;

  private final ScheduledExecutorService scheduler = Executors.newSingleThreadScheduledExecutor();

  private int totalSize;
  private final int maxSize;

  private int numberUnconfirmedTransactionsFullHash;
  private final int maxPercentageUnconfirmedTransactionsFullHash;

  final Runnable cleanupExpiredTransactions = new Runnable() {
    @Override
    public void run() {
      synchronized (internalStore) {
        final List<Transaction> expiredTransactions = getAll().stream().filter(t -> timeService.getEpochTime() > t.getExpiration()).collect(Collectors.toList());

        expiredTransactions.stream().forEach(t -> removeTransaction(t));
      }
    }
  };

  public UnconfirmedTransactionStoreImpl(TimeService timeService, PropertyService propertyService, AccountStore accountStore) {
    this.timeService = timeService;

    this.reservedBalanceCache = new ReservedBalanceCache(accountStore);

    this.maxSize = propertyService.getInt(Props.P2P_MAX_UNCONFIRMED_TRANSACTIONS);
    this.totalSize = 0;

    this.maxPercentageUnconfirmedTransactionsFullHash = propertyService.getInt(Props.P2P_MAX_PERCENTAGE_UNCONFIRMED_TRANSACTIONS_FULL_HASH_REFERENCE);
    this.numberUnconfirmedTransactionsFullHash = 0;

    internalStore = new TreeMap<>();

    scheduler.scheduleWithFixedDelay(cleanupExpiredTransactions, 1, 1, TimeUnit.MINUTES);
  }

  @Override
  public boolean put(Transaction transaction, Peer peer) throws ValidationException {
    synchronized (internalStore) {
      if(transactionIsCurrentlyInCache(transaction)) {
        if(peer != null) {
          logger.info("Transaction {}: Added fingerprint of {}", transaction.getId(), peer.getPeerAddress());
          fingerPrintsOverview.get(transaction).add(peer);
        }
      } else {
        if (transactionCanBeAddedToCache(transaction)) {
          final TransactionDuplicationResult duplicationInformation = transactionDuplicatesChecker.removeCheaperDuplicate(transaction);

          if (duplicationInformation.isDuplicate()) {
            final Transaction duplicatedTransaction = duplicationInformation.getTransaction();

            if (duplicatedTransaction != null && duplicatedTransaction != transaction) {
              logger.info("Transaction {}: Adding more expensive duplicate transaction", transaction.getId());
              removeTransaction(duplicationInformation.getTransaction());

              addTransaction(transaction, peer);

              if (totalSize > maxSize) {
                removeCheapestFirstToExpireTransaction();
              }
            } else {
              logger.info("Transaction {}: Will not add a cheaper duplicate UT", transaction.getId());
            }
          } else {
            addTransaction(transaction, peer);
            logger.debug(
                "Cache size: " + totalSize + "/" + maxSize + " Added UT " + transaction.getId() + " " + transaction.getSenderId() + " " + transaction.getAmountNQT() + " " + transaction.getFeeNQT());
            if (totalSize > maxSize) {
              removeCheapestFirstToExpireTransaction();
            }
          }

          return true;
        } else {
          logger.info("Transaction {}: Will not add UT due to duplication, or too full", transaction.getId());
        }
      }

      return false;
    }
  }

  private boolean transactionCanBeAddedToCache(Transaction transaction) {
    return transactionIsCurrentlyValid(transaction)
        && ! cacheFullAndTransactionCheaperThanAllTheRest(transaction)
        && ! tooManyTransactionsWithReferencedFullHash(transaction)
        && ! tooManyTransactionsForSlotSize(transaction);
  }

  private boolean tooManyTransactionsForSlotSize(Transaction transaction) {
    final long slotHeight = this.amountSlotForTransaction(transaction);

    return this.internalStore.containsKey(slotHeight) && this.internalStore.get(slotHeight).size() == slotHeight * 360;
  }

  private boolean tooManyTransactionsWithReferencedFullHash(Transaction transaction) {
    return ! StringUtils.isEmpty(transaction.getReferencedTransactionFullHash()) && maxPercentageUnconfirmedTransactionsFullHash <= (((numberUnconfirmedTransactionsFullHash + 1) * 100) / maxSize);
  }

  private boolean cacheFullAndTransactionCheaperThanAllTheRest(Transaction transaction) {
    return totalSize == maxSize && internalStore.firstKey() > amountSlotForTransaction(transaction);
  }

  @Override
  public Transaction get(Long transactionId) {
    synchronized (internalStore) {
      for (List<Transaction> amountSlot : internalStore.values()) {
        for (Transaction t : amountSlot) {
          if (t.getId() == transactionId) {
            return t;
          }
        }
      }

      return null;
    }
  }

  @Override
  public boolean exists(Long transactionId) {
    synchronized (internalStore) {
      return get(transactionId) != null;
    }
  }

  @Override
  public List<Transaction> getAll() {
    synchronized (internalStore) {
      final ArrayList<Transaction> flatTransactionList = new ArrayList<>();

      for (List<Transaction> amountSlot : internalStore.values()) {
        flatTransactionList.addAll(amountSlot);
      }

      return flatTransactionList;
    }
  }

  @Override
  public List<Transaction> getAllFor(Peer peer) {
    synchronized (internalStore) {
      final List<Transaction> untouchedTransactions = fingerPrintsOverview.entrySet().stream()
          .filter(e -> ! e.getValue().contains(peer))
          .map(e -> e.getKey()).collect(Collectors.toList());

      final ArrayList<Transaction> resultList = new ArrayList<>();

      long roomLeft = 175000;

      for (Transaction t : untouchedTransactions) {
        roomLeft -= t.getSize();

        if (roomLeft > 0) {
          resultList.add(t);
        } else {
          break;
        }
      }

      return resultList;
    }
  }

  @Override
  public void forEach(Consumer<Transaction> consumer) {
    synchronized (internalStore) {
      for (List<Transaction> amountSlot : internalStore.values()) {
        amountSlot.stream().forEach(consumer);
      }
    }
  }

  @Override
  public void remove(Transaction transaction) {
    synchronized (internalStore) {
      logger.debug("Removing " + transaction.getId());
      if (exists(transaction.getId())) {
        removeTransaction(transaction);
      }
    }
  }

  @Override
  public void clear() {
    synchronized (internalStore) {
      logger.debug("Clearing UTStore");
      totalSize = 0;
      internalStore.clear();
      reservedBalanceCache.clear();
      transactionDuplicatesChecker.clear();
    }
  }

  @Override
  public List<Transaction> resetAccountBalances() {
    synchronized (internalStore) {
      return reservedBalanceCache.rebuild(getAll());
    }
  }

  @Override
  public void markFingerPrintsOf(Peer peer, List<Transaction> transactions) {
    synchronized (internalStore) {
      for (Transaction transaction : transactions) {
        fingerPrintsOverview.get(transaction).add(peer);
      }
    }
  }

  private boolean transactionIsCurrentlyInCache(Transaction transaction) {
    return fingerPrintsOverview.containsKey(transaction);
  }

  private void addTransaction(Transaction transaction, Peer peer) throws ValidationException {
    this.reservedBalanceCache.reserveBalanceAndPut(transaction);

    final List<Transaction> slot = getOrCreateAmountSlotForTransaction(transaction);
    slot.add(transaction);
    totalSize++;

    fingerPrintsOverview.put(transaction, new HashSet<>());

    if(peer != null) {
      fingerPrintsOverview.get(transaction).add(peer);
    }

    logger.info("Adding Transaction {} from Peer {}", transaction.getId(), (peer == null ? "Ourself" : peer.getPeerAddress()));

    if(! StringUtils.isEmpty(transaction.getReferencedTransactionFullHash())) {
      numberUnconfirmedTransactionsFullHash++;
    }
  }

  private List<Transaction> getOrCreateAmountSlotForTransaction(Transaction transaction) {
    final long amountSlotNumber = amountSlotForTransaction(transaction);

    if (!this.internalStore.containsKey(amountSlotNumber)) {
      this.internalStore.put(amountSlotNumber, new ArrayList<>());
    }

    return this.internalStore.get(amountSlotNumber);
  }


  private long amountSlotForTransaction(Transaction transaction) {
    return transaction.getFeeNQT() / Constants.FEE_QUANT;
  }

  private void removeCheapestFirstToExpireTransaction() {
    this.internalStore.get(this.internalStore.firstKey()).stream()
        //.map(UnconfirmedTransactionTiming::getTransaction)
        .sorted(Comparator.comparingLong(Transaction::getFeeNQT).thenComparing(Transaction::getExpiration).thenComparing(Transaction::getId))
        .findFirst().ifPresent(t -> removeTransaction(t));
  }

  private boolean transactionIsCurrentlyValid(Transaction transaction) {
    if(timeService.getEpochTime() < transaction.getExpiration()) {
      return true;
    } else {
      logger.debug("Transaction {} past expiration: {}", transaction.getId(), transaction.getExpiration());
      return false;
    }
  }

  private void removeTransaction(Transaction transaction) {
    final long amountSlotNumber = amountSlotForTransaction(transaction);

    final List<Transaction> amountSlot = internalStore.get(amountSlotNumber);

    final Iterator<Transaction> transactionSlotIterator = amountSlot.iterator();

    fingerPrintsOverview.remove(transaction);

    while (transactionSlotIterator.hasNext()) {
      final Transaction utt = transactionSlotIterator.next();
      if (utt.getId() == transaction.getId()) {
        transactionSlotIterator.remove();
        transactionDuplicatesChecker.removeTransaction(transaction);
        this.reservedBalanceCache.refundBalance(transaction);
        totalSize--;

        if(! StringUtils.isEmpty(transaction.getReferencedTransactionFullHash())) {
          numberUnconfirmedTransactionsFullHash--;
        }
        return;
      }
    }

    if (amountSlot.isEmpty()) {
      this.internalStore.remove(amountSlotNumber);
    }
  }

}

File: src/main/java/brs/unconfirmedtransactions/UnconfirmedTransactionStoreImpl.kt
package brs.unconfirmedtransactions

import brs.Constants
import brs.DependencyProvider
import brs.Transaction
import brs.peer.Peer
import brs.props.Props
import brs.taskScheduler.Task
import brs.transactionduplicates.TransactionDuplicatesCheckerImpl
import brs.util.logging.safeDebug
import brs.util.logging.safeInfo
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import org.slf4j.LoggerFactory
import java.util.*

class UnconfirmedTransactionStoreImpl private constructor(private val dp: DependencyProvider) : UnconfirmedTransactionStore {
    private val reservedBalanceCache = ReservedBalanceCache(dp.accountStore)
    private val transactionDuplicatesChecker = TransactionDuplicatesCheckerImpl()

    private val fingerPrintsOverview = mutableMapOf<Transaction, MutableSet<Peer?>>()

    private val internalStore: SortedMap<Long, MutableList<Transaction>> = TreeMap()
    private val internalStoreLock = Mutex()

    override var amount: Int = 0
    private val maxSize = dp.propertyService.get(Props.P2P_MAX_UNCONFIRMED_TRANSACTIONS)

    private val maxRawUTBytesToSend = dp.propertyService.get(Props.P2P_MAX_UNCONFIRMED_TRANSACTIONS_RAW_SIZE_BYTES_TO_SEND)

    private val maxPercentageUnconfirmedTransactionsFullHash = dp.propertyService.get(Props.P2P_MAX_PERCENTAGE_UNCONFIRMED_TRANSACTIONS_FULL_HASH_REFERENCE)
    private var numberUnconfirmedTransactionsFullHash: Int = 0

    val cleanupExpiredTransactions: Task = {
        internalStoreLock.withLock {
            getAllNoLock().filter { t -> dp.timeService.epochTime > t.expiration || dp.transactionDb.hasTransaction(t.id) }
                .forEach { removeTransaction(it) }
        }
    }

    /**
     * Assumes lcoked.
     */
    private fun getAllNoLock(): List<Transaction> {
        val flatTransactionList = mutableListOf<Transaction>()

        for (amountSlot in internalStore.values) {
            flatTransactionList.addAll(amountSlot)
        }

        return flatTransactionList
    }


    override val all: List<Transaction>
        get() = runBlocking {
            internalStoreLock.withLock {
                getAllNoLock()
            }
        }

    override suspend fun put(transaction: Transaction, peer: Peer?): Boolean {
        internalStoreLock.withLock {
            if (transactionIsCurrentlyInCache(transaction)) {
                if (peer != null) {
                    logger.safeInfo { "Transaction ${transaction.id}: Added fingerprint of ${peer.peerAddress}" }
                    fingerPrintsOverview[transaction]!!.add(peer)
                }
            } else if (transactionCanBeAddedToCache(transaction)) {
                this.reservedBalanceCache.reserveBalanceAndPut(transaction)

                val duplicationInformation = transactionDuplicatesChecker.removeCheaperDuplicate(transaction)

                if (duplicationInformation.isDuplicate) {
                    val duplicatedTransaction = duplicationInformation.transaction

                    if (duplicatedTransaction != null && duplicatedTransaction !== transaction) {
                        logger.safeInfo { "Transaction ${transaction.id}: Adding more expensive duplicate transaction" }
                        removeTransaction(duplicationInformation.transaction)
                        this.reservedBalanceCache.refundBalance(duplicationInformation.transaction)

                        addTransaction(transaction, peer)

                        if (amount > maxSize) {
                            removeCheapestFirstToExpireTransaction()
                        }
                    } else {
                        logger.safeInfo { "Transaction ${transaction.id}: Will not add a cheaper duplicate UT" }
                    }
                } else {
                    addTransaction(transaction, peer)
                    if (amount % 128 == 0) {
                        logger.safeInfo { "Cache size: $amount/$maxSize added ${transaction.id} from sender ${transaction.senderId}" }
                    } else {
                        logger.safeDebug { "Cache size: $amount/$maxSize added ${transaction.id} from sender ${transaction.senderId}" }
                    }
                }

                if (amount > maxSize) {
                    removeCheapestFirstToExpireTransaction()
                }

                return true
            }

            return false
        }
    }

    /**
     * Assumes locked.
     */
    private fun getNoLock(transactionId: Long?): Transaction? {
        for (amountSlot in internalStore.values) {
            for (t in amountSlot) {
                if (t.id == transactionId) {
                    return t
                }
            }
        }
        return null
    }

    override suspend fun get(transactionId: Long?): Transaction? {
        internalStoreLock.withLock {
            return getNoLock(transactionId)
        }
    }

    override suspend fun exists(transactionId: Long?): Boolean {
        internalStoreLock.withLock {
            return getNoLock(transactionId) != null
        }
    }

    override suspend fun getAllFor(peer: Peer): List<Transaction> {
        internalStoreLock.withLock {
            var roomLeft = this.maxRawUTBytesToSend.toLong()
            return fingerPrintsOverview.entries
                    .asSequence()
                    .filter { e -> !e.value.contains(peer) }
                    .map { it.key }
                    .toList()
                    .filter {
                        roomLeft -= it.size.toLong()
                        return@filter roomLeft > 0
                    }
                    .toList()
        }
    }

    /**
     * Assumes locked.
     */
    private suspend fun removeNoLock(transaction: Transaction) {
        // Make sure that we are acting on our own copy of the transaction, as this is the one we want to remove. TODO check this
        val internalTransaction = getNoLock(transaction.id)
        if (internalTransaction != null) {
            logger.safeDebug { "Removing ${transaction.id}" }
            removeTransaction(internalTransaction)
        }
    }

    override suspend fun remove(transaction: Transaction) {
        internalStoreLock.withLock {
            removeNoLock(transaction)
        }
    }

    override suspend fun clear() {
        internalStoreLock.withLock {
            logger.safeInfo { "Clearing UTStore" }
            amount = 0
            internalStore.clear()
            reservedBalanceCache.clear()
            transactionDuplicatesChecker.clear()
        }
    }

    override suspend fun resetAccountBalances() {
        internalStoreLock.withLock {
            for (insufficientFundsTransactions in reservedBalanceCache.rebuild(getAllNoLock())) {
                removeTransaction(insufficientFundsTransactions)
            }
        }
    }

    override suspend fun markFingerPrintsOf(peer: Peer?, transactions: Collection<Transaction>) {
        internalStoreLock.withLock {
            for (transaction in transactions) {
                if (fingerPrintsOverview.containsKey(transaction)) {
                    fingerPrintsOverview[transaction]!!.add(peer)
                }
            }
        }
    }

    override suspend fun removeForgedTransactions(transactions: Collection<Transaction>) {
        internalStoreLock.withLock {
            for (t in transactions) {
                removeNoLock(t)
            }
        }
    }

    private fun transactionIsCurrentlyInCache(transaction: Transaction): Boolean {
        val amountSlot = internalStore[amountSlotForTransaction(transaction)]
        return amountSlot != null && amountSlot.any { t -> t.id == transaction.id }
    }

    private fun transactionCanBeAddedToCache(transaction: Transaction): Boolean {
        return (transactionIsCurrentlyNotExpired(transaction)
                && !cacheFullAndTransactionCheaperThanAllTheRest(transaction)
                && !tooManyTransactionsWithReferencedFullHash(transaction)
                && !tooManyTransactionsForSlotSize(transaction))
    }

    private fun tooManyTransactionsForSlotSize(transaction: Transaction): Boolean {
        val slotHeight = this.amountSlotForTransaction(transaction)

        if (this.internalStore.containsKey(slotHeight) && this.internalStore[slotHeight]!!.size.toLong() == slotHeight * 360) {
            logger.safeInfo { "Transaction ${transaction.id}: Not added because slot $slotHeight is full" }
            return true
        }

        return false
    }

    private fun tooManyTransactionsWithReferencedFullHash(transaction: Transaction): Boolean {
        if (transaction.referencedTransactionFullHash != null && maxPercentageUnconfirmedTransactionsFullHash <= (numberUnconfirmedTransactionsFullHash + 1) * 100 / maxSize) {
            logger.safeInfo { "Transaction ${transaction.id}: Not added because too many transactions with referenced full hash" }
            return true
        }

        return false
    }

    private fun cacheFullAndTransactionCheaperThanAllTheRest(transaction: Transaction): Boolean {
        if (amount == maxSize && internalStore.firstKey() > amountSlotForTransaction(transaction)) {
            logger.safeInfo { "Transaction ${transaction.id}: Not added because cache is full and transaction is cheaper than all the rest" }
            return true
        }

        return false
    }

    private fun transactionIsCurrentlyNotExpired(transaction: Transaction): Boolean {
        return if (dp.timeService.epochTime < transaction.expiration) {
            true
        } else {
            logger.safeInfo { "Transaction ${transaction.id} past expiration: ${transaction.expiration}" }
            false
        }
    }

    /**
     * Assumes locked.
     */
    private fun addTransaction(transaction: Transaction, peer: Peer?) {
        val slot = getOrCreateAmountSlotForTransaction(transaction)
        slot.add(transaction)
        amount++

        fingerPrintsOverview[transaction] = mutableSetOf()

        if (peer != null) {
            fingerPrintsOverview[transaction]!!.add(peer)
        }

        if (true) {
            if (peer == null) {
                logger.safeDebug { "Adding Transaction ${transaction.id} from ourself" }
            } else {
                logger.safeDebug { "Adding Transaction ${transaction.id} from Peer ${peer.peerAddress}" }
            }
        }

        if (transaction.referencedTransactionFullHash != null) {
            numberUnconfirmedTransactionsFullHash++
        }
    }

    private fun getOrCreateAmountSlotForTransaction(transaction: Transaction): MutableList<Transaction> {
        val amountSlotNumber = amountSlotForTransaction(transaction)

        if (!this.internalStore.containsKey(amountSlotNumber)) {
            this.internalStore[amountSlotNumber] = mutableListOf()
        }

        return this.internalStore[amountSlotNumber]!!
    }


    private fun amountSlotForTransaction(transaction: Transaction): Long {
        return transaction.feeNQT / Constants.FEE_QUANT
    }

    private fun removeCheapestFirstToExpireTransaction() {
        val cheapestFirstToExpireTransaction = this.internalStore[this.internalStore.firstKey()]!!
                .sortedWith(Comparator.comparingLong<Transaction> { it.feeNQT }
                        .thenComparing<Int> { it.expiration }
                        .thenComparing<Long> { it.id })
                .firstOrNull()
        if (cheapestFirstToExpireTransaction != null) {
            reservedBalanceCache.refundBalance(cheapestFirstToExpireTransaction)
            removeTransaction(cheapestFirstToExpireTransaction)
        }
    }

    /**
     * Assumes locked.
     */
    private fun removeTransaction(transaction: Transaction?) {
        if (transaction == null) return
        val amountSlotNumber = amountSlotForTransaction(transaction)

        val amountSlot = internalStore[amountSlotNumber]!!

        fingerPrintsOverview.remove(transaction)
        amountSlot.remove(transaction)
        amount--
        transactionDuplicatesChecker.removeTransaction(transaction)

        if (transaction.referencedTransactionFullHash != null) {
            numberUnconfirmedTransactionsFullHash--
        }

        if (amountSlot.isEmpty()) {
            this.internalStore.remove(amountSlotNumber)
        }
    }

    companion object {
        private val logger = LoggerFactory.getLogger(UnconfirmedTransactionStoreImpl::class.java)

        suspend fun new(dp: DependencyProvider): UnconfirmedTransactionStore {
            val utStore = UnconfirmedTransactionStoreImpl(dp)

            dp.taskScheduler.scheduleTaskWithDelay(utStore.cleanupExpiredTransactions, 10000, 10000)

            return utStore
        }
    }
}
